File: ConnectedComponents.java
import java.util.*;
import java.io.*;

class Graph {

    void dfs(ArrayList<ArrayList<Integer>> graph, int node, ArrayList<Boolean> visited, ArrayList<Integer> component) {
        visited.set(node, true);
        component.add(node);
        for (int neighbour : graph.get(node)) {
            if (visited.get(neighbour) != true) {
                dfs(graph, neighbour, visited, component);
            }
        }
    }

    ArrayList<ArrayList<Integer>> findConnectedComponents(ArrayList<ArrayList<Integer>> graph) {
        int n = graph.size();
        ArrayList<Boolean> visited = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            visited.add(false);
        }
        ArrayList<ArrayList<Integer>> connectedComponents = new ArrayList<>();
        for (int i = 0; i < graph.size(); i++) {
            if (visited.get(i) != true) {
                ArrayList<Integer> component = new ArrayList<>();
                dfs(graph, i, visited, component);
                connectedComponents.add(component);
            }
        }

        return connectedComponents;
    }
}

public class ConnectedComponents {
    public static void main(String[] args) {
        // System.out.println("Hello World!!!");
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();
        ArrayList<ArrayList<Integer>> graph = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }
        for (int i = 0; i < m; i++) {
            int u = sc.nextInt();
            int v = sc.nextInt();
            graph.get(u).add(v);
            graph.get(v).add(u);
        }
        Graph obj = new Graph();

        ArrayList<ArrayList<Integer>> connectedComponents = obj.findConnectedComponents(graph);
        int count = 1;
        for (ArrayList<Integer> component : connectedComponents) {
            System.out.print(count++ + ": ");
            for (int node : component) {
                System.out.print(node + " ");
            }
            System.out.println();
        }
    }
}

File: ConnectedComponents.cpp
#include <algorithm>
#include <cmath>
#include <iostream>
#include <list>
#include <queue>
#include <set>
#include <string>
#include <vector>

using namespace std;

typedef long long ll;
#define ln "\n"
#define mp make_pair
#define pb push_back
#define f first
#define s second
#define INF 2e18

void dfs(vector<vector<int>> &graph, int node, vector<bool> &visited,
         vector<int> &component) {
  visited[node] = true;
  component.push_back(node);
  for (int neighbor : graph[node]) {
    if (!visited[neighbor]) {
      dfs(graph, neighbor, visited, component);
    }
  }
}

vector<vector<int>> findConnectedComponents(vector<vector<int>> &graph) {
  vector<bool> visited(graph.size(), false);
  vector<vector<int>> connectedComponents;
  for (int i = 0; i < graph.size(); i++) {
    if (!visited[i]) {
      vector<int> component;
      dfs(graph, i, visited, component);
      connectedComponents.push_back(component);
    }
  }
  return connectedComponents;
}

int main() {
  // cout<<"Hello world !";
  int n, m;
  cin >> n >> m;
  vector<vector<int>> graph(n);
  for (int i = 0; i < m; i++) {
    int u, v;
    cin >> u >> v;
    graph[u].push_back(v);
    graph[v].push_back(u);
  }
  vector<vector<int>> connectedComponents = findConnectedComponents(graph);
  int count = 1;
  for (const vector<int> &component : connectedComponents) {
    cout << count++ << ": ";
    for (int node : component) {
      cout << node << " ";
    }
    cout << endl;
  }
}

// 10 6
// 0 1
// 1 2
// 3 4
// 4 5
// 6 7
// 8 9